MemberService 단위테스트작성 #33

nahokyun opened the pull request:
> ### 추가된점
- MemberService의 회원가입, 아이디 존재여부 체크, login 성공여부 체크를 하는 단위 테스트 코드를 추가하였습니다.
## MemberServiceTest.java
### 회원가입
```java
    @Test
    public void addMember_회원가입() {
        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"20","test");

        doNothing().when(memberMapper).save(any(Member.class));
        memberService.addMember(memberDTO);

        verify(memberMapper, times(1)).save(any(Member.class));
    }
```
- memberMapper에서  save함수를 1번 호출 했었는지 확인합니다.
### 중복 아이디체크
```java
    @Test
    void checkLoginId_존재할때() {
        doReturn(true).when(memberMapper).findByLoginId(testId);

        assertThatThrownBy(()->memberService.checkLoginId(testId))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void checkLoginId_존재하지않을때() {
        doReturn(false).when(memberMapper).findByLoginId(testId);

        memberService.checkLoginId(testId);

        verify(memberMapper,times(1)).findByLoginId(anyString());
    }
```
- 이미 존재하는 아이디를 입력했을 경우 IllegalArgumentException이 발생하며 오류 메시지를 출력하기때문에 이를 catch하여 구현하였습니다.
- 존재하지 않을 경우에는 memberMapper에서 findByLoginId가 1번 호출되었는지 확인합니다.
### 로그인
```java
    @Test
    void logIn_성공할때() throws LoginException {
        Member member=member();
        Member encodedMember=encodedMember();

        doReturn(Optional.of(encodedMember)).when(memberMapper).getMemberForLogIn(testId);
        doNothing().when(sessionManager).storeLoginId(1L);

        memberService.logIn(testId,member.getLoginPassword());

        verify(sessionManager,times(1)).storeLoginId(anyLong());
    }

    @Test
    void logIn_성공못할때(){
        Member member=member();

        doReturn(Optional.empty()).when(memberMapper).getMemberForLogIn(testId);

        assertThatThrownBy(()->memberService.logIn(testId,member.getLoginPassword()))
                .isInstanceOf(LoginException.class);

        verify(memberMapper,times(1)).getMemberForLogIn(anyString());
        verify(sessionManager,times(0)).storeLoginId(anyLong());
    }

```
- login이 성공할때에는 입력 password와 암호화된 password가 서로 match가 되는지 확인후 sessionManager에서 memberId를 저장하는 storeLoginId를 1번 호출했는지 확인합니다.
- login이 실패할때에는 LoginException과 오류메시지가 출력되는데 이를 catch하고 getMemberForLogIn과 storeLoginId가 각각 1번 0번 호출되었는지 확인합니다.

nahokyun committed:
MemberService 단위테스트작성 #33

MemberService의 회원가입, 아이디 존재여부 체크, login 성공여부 체크를 하는 테스트 코드를 추가하였습니다.

Jeonminje commented on src/test/java/study/movieservice/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Test
+    void checkLoginId_존재할때() {
+        doReturn(true).when(memberMapper).findByLoginId(testId);
```
매퍼의 findByLoginId() 함수를 테스트 하신걸까요??

nahokyun replied:
해당 메소드의 로직이 제대로 동작하는지에대한 테스트를 진행하였습니다. 이 테스트는 memberMapper의 findByLoginId를 testId를 넘겨주며 호출하였을때 true리턴함으로써 true를 리턴받았을때 checkLoginId가 성공적으로 IllegalArgumentException을 throw를 하는지를 체크한것입니다. 단위테스트의 목적에 초점을 맞춘다면 findByLoginId함수의 테스트는 memberMapper에 대한 단위테스트코드를 작성할때 넣는것이 맞다고 생각합니다!

sunwootest replied:
MemberService는 Mock객체가 아니기 때문에 이렇게 mapper로 가정처리 하신건 잘하셨습니다 :)

Jeonminje commented on src/test/java/study/movieservice/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,104 @@
+@ExtendWith(MockitoExtension.class)
+class MemberServiceTest {
+
+    @Mock
```
이 어노테이션을 사용한 이유가 무엇일까요??

nahokyun replied:
ExtendWith 어노테이션의 경우 단위테스트를 진행할때 MockitoExtension.class의 기능을 사용을 위해 확장하기 위해서 사용한 것이고 Mock 어노테이션의 경우 가짜객체임을 명시하고 InjectMocks어노테이션을 사용한 meberService에 주입하기 위하여 사용하였습니다.

Jeonminje reviewed (commented):
제가 느낀 부분이 맞다면, testId가 DB에 저장되어있다는 가정 하에 테스트가 진행되는거 같습니다! 
만약, DB에 testId가 저장되어있지 않은 상태에서 테스트코드들을 한꺼번에 실행한다면,
실행되는 함수의 순서가 보장되지 않기때문에 addMember_회원가입() 메서드가 먼저실행되지 않는다면, 다른 함수들이 fail을 뱉을거 같습니다!

sunwootest reviewed (commented):
단위테스트 작성은 잘하셨습니다. 테스트 내부 구현자체보다는 어떤걸 테스트해야할지에 집중하시면 좋을 것 같습니다. 예를들어 회원가입이 어떤 경우에 실패해야되는지 케이스를 작성하시고 그 케이스들에 대해서 테스트를 진행하면 좋을 것 같습니다. 그리고 테스트 메소드 명이 알아보기 좀 힘든 것 같습니다 .길어도 좋으니 최대한 메소드명으로 파악 가능하게 해주시면 좋을 것 같습니다 . DisplayName 어노테이션을 사용해도 좋습니다 :)

nahokyun committed:
MemberService 단위테스트작성 #33

메소드명 구체적으로 변경

nahokyun committed:
Member정보관련 유효성 검사 추가#39

validation사용을위해 org.springframework.boot:spring-boot-starter-validation를 추가하였습니다.

회원가입에 실패했을때의 message를 추가하였습니다.

validation에 걸렸을때 발생하는 exception을 handling하였습니다.

nahokyun committed:
MemberService 단위테스트 작성#33

DisplayName을 사용하여 test 상황작성

nahokyun committed:
Member정보관련 유효성 검사 추가#39

loginId와 nickname에 대한 유효성 조건을 추가하였습니다.

nahokyun committed:
Member정보관련 유효성 검사 추가#39

MemberService 클래스에 Validated어노테이션과 addMember 메소드에 @Valid를 사용함으로 service단계에서 MemberDTO에대한 유효성 검사를 진행할수 있도록 추가하였습니다.

nahokyun committed:
Merge branch 'Validation-memberDTO#39' into MemberService-Test

nahokyun commented:
이중으로 memberDTO의 유효성검사를 진행할때 테스트 코드에 mockMVC까지 사용하는것은 MemberService 단위테스트 작성에 맞지 않는경우일까요? MemberService layer에서 @Validated와 @Valid를 사용하여 실제로 실행을 통해 유효성검사가 진행되는것을 확인은 하였는데 테스트 코드에서는 @Valid가 제대로 적용되지 않는것같아 mockMVC외에 검증할만한 방법이 있을까요?

sunwootest commented:
> 이중으로 memberDTO의 유효성검사를 진행할때 테스트 코드에 mockMVC까지 사용하는것은 MemberService 단위테스트 작성에 맞지 않는경우일까요? MemberService layer에서 @Validated와 @Valid를 사용하여 실제로 실행을 통해 유효성검사가 진행되는것을 확인은 하였는데 테스트 코드에서는 @Valid가 제대로 적용되지 않는것같아 mockMVC외에 검증할만한 방법이 있을까요?

이건 통합테스트에 가깝긴 합니다. spring에서 유효성 검사가 어떤 원리로 이루어지는지 이해하면 테스트를 어떻게 작성해야할지 감이 올겁니다 :)

nahokyun committed:
MemberService 단위테스트 작성#33

@valid를 사용한 유효성 검사가 테스트환경에서 잘 돌아가지않음을 체크했고 이에따른 방안찾아야함

nahokyun committed:
MemberService 단위테스트작성 #33

validator를 통한 유효성 검사 로직을 추가하였습니다.

nahokyun commented:
> ### 변경된점
- validator를 사용하여 addMember와 addMember_FAILED의 유효성 검사를 추가하였습니다.
## MemberService.java
### 변경 전 
```java
   @Test
    @DisplayName("회원가입 성공할때")
    public void addMember() {
        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"20","test");

        doNothing().when(memberMapper).save(any(Member.class));
        memberService.addMember(memberDTO);

        verify(memberMapper, times(1)).save(any(Member.class));
    }

    @Test
    @DisplayName("회원가입 실패할때_컨트롤러에서 전달되는 memberDTO가 정상이 아닐때")
    public void addMember_FAILED(){

        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"","1");

        assertThatThrownBy(()->memberService.addMember(memberDTO))
                .isInstanceOf(ConstraintViolationException.class);
    }

```
### 변경 후 
```java
   @Test
   @DisplayName("회원가입 성공할때")
   public void addMember() {
        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"20","test");

        doNothing().when(memberMapper).save(any(Member.class));
        memberService.addMember(memberDTO);

        Set<ConstraintViolation<MemberDTO>> violations = validator.validate(memberDTO);

        verify(memberMapper, times(1)).save(any(Member.class));
        assertThat(violations.size()).isEqualTo(0);
    }

    @Test
    @DisplayName("회원가입 실패할때_컨트롤러에서 전달되는 memberDTO가 정상이 아닐때")
    public void addMember_FAILED(){
        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"2","1");

        Set<ConstraintViolation<MemberDTO>> violations = validator.validate(memberDTO);

        assertThat(violations.size()).isEqualTo(2);
    }

```

sunwootest commented on src/test/java/study/movieservice/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,134 @@
+
+        verify(memberMapper, times(1)).save(any(Member.class));
+        assertThat(violations.size()).isEqualTo(0);
+    }
```
any를 이용하여 Member.class가 들어간지 확인하면 정확한 회원가입 성공 테스트일까요? any를 왜 사용했는지는 모르겠지만 동등성 동일성에 관해서 생각해보면 더 좋은 테스트코드를 작성할 수 있을겁니다.

nahokyun committed:
MemberService 단위 테스트작성#33

hashcode()와 equals()를 override하여 논리적 값이 같으면 같은 hashcode를 return 하도록 변경하여 test를 진행하였습니다.

nahokyun commented:
> ### 변경된점
- static method인 BCrypt.hashpw값 고정을 위해 build.gradle에 testimplementation 'org.mockito:mockito-inline:3.6.0'를 추가하였습니다.
- Member 클래스의 hashcode()와 equals()를 Override하여 논리적인 값이 같을경우 같은 hashcode 값을 갖도록 변경하였습니다.
## MemberServiceTest.java
### 변경 전 
```java
    @Test
    @DisplayName("회원가입 성공할때")
    public void addMember() {
        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"20","test");

        doNothing().when(memberMapper).save(any(Member.class));
        memberService.addMember(memberDTO);

        Set<ConstraintViolation<MemberDTO>> violations = validator.validate(memberDTO);

        verify(memberMapper, times(1)).save(any(Member.class));
        assertThat(violations.size()).isEqualTo(0);
    }

```
### 변경 후 
```java
    @Test
    @DisplayName("회원가입 성공할때_컨트롤러에서 전달되는 memberDTO가 정상일 때")
    void addMember_SUCCESS() {

        MemberDTO memberDTO=new MemberDTO(testEmail,testId,"20",testNickname);
        Member member=member();
        MockedStatic<BCrypt> bCrypt= mockStatic(BCrypt.class);

        doNothing().when(memberMapper).save(any(Member.class));
        when(BCrypt.hashpw("20",BCrypt.gensalt())).thenReturn("20");

        memberService.addMember(memberDTO);

        Set<ConstraintViolation<MemberDTO>> violations = validator.validate(memberDTO);

        verify(memberMapper, times(1)).save(member);
        assertThat(violations.size()).isEqualTo(0);

        bCrypt.close();
    }

```
